### PR TITLE
Add a Tokenizer to Azure OpenAI

### PR DESCRIPTION
**Pull request by agoncal**
Today the Azure OpenAI module does not have its own Tokenizer. It actually uses the OpenAI one during its test phase (`langchain4j-open-ai` dependency in `test` scope).

This PR is for Azure OpenAI to have its own Tokenizer. Not just for test purposes (this PR removes the dependency to `langchain4j-open-ai`) but also Azure OpenAI can be used for RAG without the need of the `OpenAITokenizer`.

For that I had to change all the `ModelName` classes (`AzureOpenAiChatModelName`, `AzureOpenAiEmbeddingModelName`...) to add the model name (eg. `gpt-35-turbo-0301`), the model type (eg. `gpt-3.5-turbo`, notice the dot `.` in `3.5`) and version.

The `AzureOpenAiTokenizer` is a copy/paste from `OpenAiTokenizer`.

All `AzureOpenAiTokenizerTest` tests pass.

I need some help with `AzureOpenAiTokenizerIT`.  It uses the `ChatCompletionModel modelName` which has dots `.` in the names 

```
dev.ai4j.openai4j.chat;

public enum ChatCompletionModel {

    GPT_3_5_TURBO("gpt-3.5-turbo"),
```

So the `deploymentName` is not valid for Azure OpenAI. I am incline to use `AzureOpenAiChatModelName` instead of `ChatCompletionModel` but just want to know if it's the right approach

```
    void should_count_tokens_in_messages(List<ChatMessage> messages, ChatCompletionModel modelName) {

        // given
        AzureOpenAiChatModel model = AzureOpenAiChatModel.builder()
                .endpoint(System.getenv("AZURE_OPENAI_ENDPOINT"))
                .apiKey(System.getenv("AZURE_OPENAI_KEY"))
                .deploymentName(modelName.toString())
                .maxTokens(1) // we don't need outputs, let's not waste tokens
                .logRequestsAndResponses(true)
                .build();
}
```

## General checklist

- [x] There are no breaking changes
- [x] I have added unit and integration tests for my change
- [x] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [ ] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)

### CODE REVIEW

**Comment by agoncal** [comment]
Pinging @jdubois for a double check

File: langchain4j-azure-open-ai/pom.xml [hunk 1]
```diff
@@ -21,6 +21,11 @@
             <artifactId>langchain4j-core</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>dev.ai4j</groupId>
+            <artifactId>openai4j</artifactId>
```
Comment (langchain4j) [inline langchain4j-azure-open-ai/pom.xml]: can probably be removed now?

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiChatModelName.java [hunk 2]
```diff
@@ -6,37 +6,60 @@
...
+        this.modelVersion = null;
+    }
+
+    AzureOpenAiChatModelName(String modelName, String modelType, String modelVersion) {
+        this.modelName = modelName;
+        this.modelType = modelType;
+        this.modelVersion = modelVersion;
+    }
+
+    public String getModelName() {
```
Comment (langchain4j) [inline langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiChatModelName.java]:
```suggestion
    public String modelName() {
```
Please drop "get" from getters (in all classes), we follow modern record-style naming convention

**Comment by langchain4j** [comment]
@agoncal thanks a lot!

> I am incline to use AzureOpenAiChatModelName instead of ChatCompletionModel but just want to know if it's the right approach

Sounds good.

BTW, `OpenAiTokenizer`/`OpenAiTokenizerIT` are probably a bit outdated.
We need to revisit them again and check if is really needed any maybe they can be simplified.

Why `OpenAiTokenizer` was created in the first place and how it can be helpful:
- calculate tokens in streaming mode. For a long time OpenAI did not provide token usage information when streaming, but now it does, so it becomes less important.
- estimate request cost before doing the request. Not sure how needed this feature is right now, I don't have any feedback about it.
- split text for RAG. As context sizes increase, it becomes less important to split documents and to be very precise about token calculation in this case. 1 word/symbol = N tokens might be a reasonable approximation going forward.

**Comment by jdubois** [comment]
Wow @agoncal I wanted to code this too, but it seemed a ton of work, so happy you did this! It seems complicated to have a correct estimation here, but as commented by @langchain4j it seems less important today.

**Comment by agoncal** [comment]
@langchain4j I changed the model names so it uses the Azure ones. But this is difficult to test due to the combination of models. I only managed to have one test pass with GPT 3.5 Turbo. Don't know if it's enough. But as you said, Tokenizers are becoming less important.

Let me know what you think.

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiTokenizer.java [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+ * There are integration tests in place that are making sure that the calculations here are very close to that of OpenAI.
+ */
+public class AzureOpenAiTokenizer implements Tokenizer {
+
+    private final String modelName;
+    private final Optional<Encoding> encoding;
+
+    /**
+     * Creates an instance of the {@code AzureOpenAiTokenizer} for the "gpt-3.5-turbo" model.
+     * It should be suitable for all current OpenAI models, as they all use the same cl100k_base encoding.
```
Comment (langchain4j) [inline langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiTokenizer.java]: Not true any more, gpt-4o uses 200k tokens now

**Comment by langchain4j** [comment]
@agoncal depends on what this tokenizer is going to be used for, a lot of existing complexity can be removed.

Or we can keep it as it is for the moment and if I will be updating OpenAiTokenizer/tests, I can update this one as well.

File: langchain4j-azure-open-ai/pom.xml (hunk 1 above)
Comment (agoncal) [reply langchain4j-azure-open-ai/pom.xml]: `AzureOpenAiTokenizerTest` uses `openai4j` in the following methods:

```
    @ParameterizedTest
    @EnumSource(ChatCompletionModel.class)
    void should_support_all_chat_models_from_openai4j(ChatCompletionModel model)

    @ParameterizedTest
    @EnumSource(EmbeddingModel.class)
    void should_support_all_embedding_models_from_openai4j(EmbeddingModel model) {

    @ParameterizedTest
    @EnumSource(CompletionModel.class)
    void should_support_all_language_models_from_openai4j(CompletionModel model) {
```

But I don't know if that makes sense anymore as we have:

```
    @ParameterizedTest
    @EnumSource(AzureOpenAiChatModelName.class)
    void should_support_all_chat_models(AzureOpenAiChatModelName modelName) {
```

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiChatModelName.java (hunk 2 above)
Comment (agoncal) [reply langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiChatModelName.java]: Done in https://github.com/langchain4j/langchain4j/pull/1222/commits/594231b7c2976aa9a858ad6ca044cfe63d6d5272

File: langchain4j-azure-open-ai/pom.xml (hunk 1 above)
Comment (langchain4j) [reply langchain4j-azure-open-ai/pom.xml]: Ah, I see. All `from_openai4j` tests can be removed as azure module does not rely on openai4j lib

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiTokenizer.java (hunk 3 above)
Comment (agoncal) [reply langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiTokenizer.java]: Done https://github.com/langchain4j/langchain4j/pull/1222/commits/dbee1dd08af94844ab00c59ab522742995b4ca88

File: langchain4j-azure-open-ai/pom.xml (hunk 1 above)
Comment (agoncal) [reply langchain4j-azure-open-ai/pom.xml]: Done https://github.com/langchain4j/langchain4j/pull/1222/commits/8c13cb36fa8cdc71a900177d3d5d78d6362c0291

**Comment by agoncal** [comment]
@langchain4j yes, it's a copy/paste from `OpenAiTokenizer`, so if `OpenAiTokenizer` is updated, then we will update this one. It's just so Azure OpenAI does not depend on OpenAI and can have its own tokenizer, mostly for RAG purpose.

**Review by ghost — APPROVED** [review]
@agoncal thank you!

**Pull request merged**